core/src/test/scala/org/typelevel/paiges/PaigesScalacheckTest.scala

```diff
@@ -254,6 +235,48 @@ class PaigesScalacheckTest extends FunSuite {
     }
   }
 
+  test("Union invariant: `a.flatten == b.flatten`") {
+    forAll { d: Doc.Union => assert(d.a.flatten === d.b.flatten) }
+  }
+
+  test("Union invariant: `a != b`") {
+    forAll { d: Doc.Union => assert(d.a !== d.b) }
+  }
+
+  def unionRightAssocInvariant(d: Doc): Assertion = {
+    import Doc._
+    d match {
+      case Empty | Text(_) | Line(_) => assert(true)
+      case Concat(Concat(_, _), _) => assert(false, s"Left-associative Concat: ${d}")
+      case Concat(a, b) =>
+        unionRightAssocInvariant(a)
+        unionRightAssocInvariant(b)
+      case Union(a, _) => unionRightAssocInvariant(a)
+      case LazyDoc(f) => unionRightAssocInvariant(f.evaluated)
+      case Align(d) => unionRightAssocInvariant(d)
```

Comment:
I initially left `Align` and `Nest` out of the test, and they still passed.  This suggests perhaps that our `Doc` generator doesn't produce them readily enough.

Reply:
Adding `fill` to the Union generator broke the invariants, so I'm going to close this temporarily while I figure out what's wrong.